hepcrawl/spiders/edp_spider.py

```diff
@@ -0,0 +1,430 @@
+                    meta=ftp_params,
+                    callback=self.handle_package_ftp
+                )
+
+    def handle_package_ftp(self, response):
+        """Handle remote packages and yield every XML found."""
+        self.log("Visited %s" % response.url)
+        zip_filepath = response.body
+        zip_target_folder, _ = os.path.splitext(zip_filepath)
+        if "tar" in zip_target_folder:
+            zip_target_folder, _ = os.path.splitext(zip_target_folder)
+        xml_files = self.untar_files(zip_filepath, zip_target_folder)
+        for xml_file in xml_files:
+            yield Request(
+                "file://{0}".format(xml_file),
+                meta={"package_path": zip_filepath}
+            )
+
+    def handle_package_file(self, response):
+        """Handle a local package and yield every XML found."""
+        zip_filepath = urlparse.urlsplit(response.url).path
+        zip_target_folder, _ = os.path.splitext(zip_filepath)
+        if "tar" in zip_target_folder:
+            zip_target_folder, _ = os.path.splitext(zip_target_folder)
+        xml_files = self.untar_files(zip_filepath, zip_target_folder)
+        for xml_file in xml_files:
+            request = Request(
+                "file://{0}".format(xml_file),
+                meta={"package_path": zip_filepath}
+            )
+            if "xml_rich" in xml_file:
+                request.meta["rich"] = True
+                self.itertag = "EDPSArticle"
+            yield request
+
+    @staticmethod
+    def untar_files(zip_filepath, target_folder, flatten=False):
+        """Unpack the tar.gz or tar.bz2 package and return XML file paths."""
+        xml_files = []
+        with tarfile.open(zip_filepath) as tar:
+            for filename in tar.getmembers():
+                if filename.path.endswith(".xml"):
+                    if flatten:
+                        filename.name = os.path.basename(filename.name)
+                    absolute_path = os.path.join(target_folder, filename.path)
+                    if not os.path.exists(absolute_path):
+                        tar.extract(filename, path=target_folder)
+                    xml_files.append(absolute_path)
+
+        return xml_files
+
+    def parse_node(self, response, node):
+        """Parse the XML file and yield a request to scrape for the PDF."""
+        node.remove_namespaces()
+        if response.meta.get("rich"):
+            article_type = node.xpath('./ArticleID/@Type').extract_first()
+            dois = node.xpath('.//DOI/text()').extract()
+        else:
+            article_type = node.xpath('@article-type').extract_first()
+            dois = node.xpath(
+                ".//article-id[@pub-id-type='doi']/text()").extract()
+
+        self.log("Got article_type {0}".format(article_type))
+        if article_type is None or article_type not in self.allowed_article_types:
+            # Filter out non-interesting article types
+            yield None
```

Comment:
I'm not sure if this is what you actually want, `yield` is not a `return`, it will not exit the function, on next call, it will continue from where it left, for example:

```python
def a():
    for i in range(10):
        yield i
        yield i+100
```

will show:

```python
In [3]: [ j for j in a() ]
Out[3]: 
[0,
 100,
 1,
 101,
 2,
 102,
 3,
 103,
 4,
 104,
 5,
 105,
 6,
 106,
 7,
 107,
 8,
 108,
 9,
 109]
```
so both, the first `yield` (`i`) and the second (`i+100`) are actually returned.